### DISCUSSION TITLE
tinyusb with STM32H563 (Device CDC in particular)

### DISCUSSION THREAD

**Discussion by Lotharyx** [hathach/tinyusb#2275] [Q&A]
Looks like this falls into the "unimplemented variant" category...  Unfortunately I am pretty un-knowledgeable about low-level USB implementations.  STM32 doesn't even provide a HAL version for this family (at least, not yet; or at least, not one compatible with FreeRTOS).

All I need is device-mode CDC (Virtual COM Port) so I can send bytes back and forth with a host computer.  Would anybody be willing to give a newb a bit of patient guidance?  "Go read this" is welcome advice.

Also, I noticed this comment about assumptions at the top of `dcd_stm32_fsdev.c`:

> - You are not using CAN (it must share the packet buffer)

Erm...well...my application does [will] use CAN.  So does that mean I can't use TinyUSB?

**Comment by Lotharyx** [comment]
Poking around a bit more, it seems like a lot of the differences among STM32 families boils down to #including the proper family header and setting four #defines.  Can I just create an `OPT_MCU_STM32H5` macro in `tusb_option.h` and update `dwc2_stm32.h` and `dcd_synopsys.c` with a block like this?

```
#elif CFG_TUSB_MCU == OPT_MCU_STM32H5
  #include "stm32h5xx.h"
	#define EP_MAX_FS 8
	#define EP_FIFO_SIZE_FS 4096
	#define EP_MAX_HS 8
	#define EP_FIFO_SIZE_HS 4096
```

That number `8` appears in some STM example code.  Not sure what the fifo sizes should be; I'll dig through datasheets tomorrow...

**Comment by KlemenDEV** [comment]
@Lotharyx, did you get it to work? What driver did you end-up using dwc2 or stm32_fsdev?

**Comment by Lotharyx** [comment]
The story hasn’t advanced.  I have code that compiles, but I don’t have hardware to test it on yet.

**Comment by KlemenDEV** [comment]
Thanks for the reply. I have also found this issue and left my own findings there: https://github.com/hathach/tinyusb/issues/2203

**Comment by Lotharyx** [comment]
Looks like you've made it a lot farther than I have.  I finally have hardware (and time!) to test...  Joining in over on your discussion.
